### ISSUE TITLE
Recursive async generators being interpreted as a Coroutine

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#4478]
A simple reproduction example:

```python
async def foo(n: int):
    for i in range(n):
        yield i

    reveal_type(foo)
    async for i in foo(n - 1):
        yield i


reveal_type(foo)

async def main():
    async for i in foo(10):
        pass
```
Pyright output:
```
❯ pyright test.py
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.289
/tmp/test.py
  /tmp/test.py:6:20 - error: "Coroutine[Any, Any, Unknown]" is not iterable
    "__aiter__" method not defined (reportGeneralTypeIssues)
  /tmp/test.py:5:17 - information: Type of "foo" is "(n: int) -> AsyncGenerator[int | Unknown, None]"
  /tmp/test.py:10:13 - information: Type of "foo" is "(n: int) -> AsyncGenerator[int | Unknown, None]"
1 error, 0 warnings, 2 informations 
Completed in 0.696sec
```

Even though it knows that foo is an `AsyncGenerator`, when used in an `async for` loop it thinks it is a coroutine

**Comment by bellini666** [comment]
Note: If I type the return value of foo to `async def foo(n: int) -> AsyncGenerator[int]:` the error is gone
